File: 0879-profitable-schemes/0879-profitable-schemes.cpp
class Solution {
public:
    int dp[102][102][102];
    int mod=1e9+7;
    int recur(int idx,int curr_profit,int n,int minProfit,vector<int>& group, vector<int>& profit){
        if(idx==profit.size()){
            if(curr_profit>=minProfit && n>=0)return 1;
            return 0;
        }
        else if(n<=0){
            if(n==0 && curr_profit>=minProfit)return 1;
            return 0;
        } 
        if(dp[idx][curr_profit][n]!=-1)return dp[idx][curr_profit][n];
        
        int notConsidering=recur(idx+1,curr_profit,n,minProfit,group,profit);
        int considering=recur(idx+1,min(curr_profit+profit[idx],minProfit),n-group[idx],minProfit,group,profit);
        return dp[idx][curr_profit][n]= (notConsidering%mod + considering%mod)%mod;
        
    }
    
    int profitableSchemes(int n, int minProfit, vector<int>& group, vector<int>& profit) {
        memset(dp,-1,sizeof(dp));
        return recur(0,0,n,minProfit,group,profit);
    }
};

File: 0879-profitable-schemes/0879-profitable-schemes.java
class Solution {
    int mod=1_000_000_007;
    long [][][]dp;
    public long memo(int idx,int currProfit,int n,int minProfit,int[]group,int[]profit){
        
        if(idx==profit.length || n<=0){
            if(currProfit>=minProfit && n>=0)return 1;
            return 0;
        }
        if(dp[idx][n][currProfit]!=-1)return dp[idx][n][currProfit];
        
        long notConsidering=memo(idx+1,currProfit,n,minProfit,group,profit);
        long Considering=memo(idx+1,Math.min(currProfit+profit[idx],minProfit),n-group[idx],minProfit,group,profit);
        
        return dp[idx][n][currProfit]= (notConsidering+Considering)%mod;
        
    }
    
    public int profitableSchemes(int n, int minProfit, int[] group, int[] profit) {
        dp=new long[102][102][102];
        for(long [][]i:dp){
            for(long []j:i){
                Arrays.fill(j,-1);
            }
        }
        return (int)memo(0,0,n,minProfit,group,profit);
    }
}